### PR TITLE
Add `Paths` type

### LINKED ISSUE #213: Feature Request: add `Paths` type
```ts
//  by @金鑫

type Ensure<X, Y> = X extends Y ? X : never;
type SafeKey<T> = T extends string | number ? `${T}` : never;
type SafeValue<O, K> = K extends keyof O ? O[K] : never;

type Paths<O, P = ""> = O extends object
  ? {
      [K in keyof O]:
        | (P extends "" ? K : `${SafeKey<P>}.${SafeKey<K>}`)
        | Paths<O[K], P extends "" ? K : `${SafeKey<P>}.${SafeKey<K>}`>;
    }[keyof O]
  : never;

type Get<O, P extends Paths<O>> = P extends `${infer X}.${infer XS}`
  ? Get<SafeValue<O, X>, Ensure<XS, Paths<SafeValue<O, X>>>>
  : SafeValue<O, P>;
```

These types help us generate all possible paths of an object, which help build library like `lodash.get` and could be considered as companions of the already exists `Get`.

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/213">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/213/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by Emiyaaaaa**
- Close #213 
 - Close #300

### CODE REVIEW

File: source/paths.d.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		};
+	};
+};
+
+type ProjectPaths = Paths<Project>;
+//=> "fileName" | "folder" | "folder.subfolder" | "folder.subfolder.fileName"
+
+declare function open<Path extends ProjectPaths>(path: Path): void;
+
+open('fileName'); // Pass
```
Comment (sindresorhus) [inline source/paths.d.ts:30]: Show an example with array too.

File: source/paths.d.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	fileName: string;
+	folder: {
+		subfolder: {
+			fileName: string;
+		};
+	};
+};
+
+type ProjectPaths = Paths<Project>;
+//=> "fileName" | "folder" | "folder.subfolder" | "folder.subfolder.fileName"
```
Comment (sindresorhus) [inline source/paths.d.ts:26]: single-quotes

File: source/paths.d.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+Generate a union of all the possible paths in given object.
+
+Use-case: You want a type-safe way to access deeply nested properties in an object.
+
+@example
+```
+import type {Paths} from 'type-fest';
+
+type Project = {
+	fileName: string;
```
Comment (sindresorhus) [inline source/paths.d.ts:17]:
```suggestion
	filename: string;
```

**Comment by sindresorhus** [comment]
It was suggested in the issue that the `Get` type could use this. Do you think so?

File: readme.md [hunk 4]
```diff
@@ -177,6 +177,7 @@ Click the type names for complete docs.
 - [`ArrayIndices`](source/array-indices.d.ts) - Provides valid indices for a constant array or tuple.
 - [`ArrayValues`](source/array-values.d.ts) - Provides all values for a constant array or tuple.
 - [`SetFieldType`](source/set-field-type.d.ts) - Create a type that changes the type of the given keys.
+- [`Paths`](source/paths.d.ts) - Generate a union of all the possible paths in given object.
```
Comment (sindresorhus) [inline readme.md:180]:
```suggestion
- [`Paths`](source/paths.d.ts) - Generate a union of all possible paths to properties in the given object.
```

File: source/paths.d.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import type {EmptyObject} from './empty-object';
+import type {UnknownArray} from './unknown-array';
+import type {UnknownRecord} from './unknown-record';
+
+type ToString<T> = T extends string | number ? `${T}` : '';
+
+/**
+Generate a union of all the possible paths in given object.
+
```
Comment (sindresorhus) [inline source/paths.d.ts:9]: Should mention that it also works for arrays.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by Emiyaaaaa** [comment]
> It was suggested in the issue that the `Get` type could use this. Do you think so?

`Get` accept 'string' and 'string[]':
1. argument 'string' is like `hits.hits[0]._source.name`, but `Paths` current result like `hits.hits.0._source.name`
I'm having some trouble:
Is result of `Paths` wrong? 
or should add a option to choose the handling of array? 
or path in `Get` and `Paths` just are not same thing?
3. argument 'string[]' is like `['hits','hits', '0', '_source', 'name']`, it can replace with `Split<Paths<OBJECT>, '.'>`

And this is a broken change if replace type directly, so we should think about backward compatibility, example we can use `LiteralUnion<Paths<T>, string>` to replace `string` in `Get`

**Comment by Emiyaaaaa** [comment]
>   Is result of `Paths` wrong?
>   or should add a option to choose the handling of array?

Should resolve this question in current PR, need your help~ @sindresorhus

**Comment by sindresorhus** [comment]
I forgot that `Get` accepts `[0]`. Then it doesn't make sense to use `Paths`. We can consider adding an option for `Paths` to also generate `[0]` paths, but that shouldn't be part of this pull request and we should discuss it in an issue first.

**Comment by Emiyaaaaa** [comment]
Changes: 
 - resolved review
 - add support for tuple with spread item
 - now will return `'0' | 0` for `Paths<[string]>`
 - now will return `never` for the type witch haven't any paths, like empty object and empty array.

**Comment by sindresorhus** [comment]
Nice work 🎉

File: source/paths.d.ts [hunk 6]
```diff
@@ -92,11 +95,11 @@ export type InternalPaths<_T extends UnknownRecord | UnknownArray, T = Required<
 		? never
 		: {
 			[Key in keyof T]:
-			Key extends string | number // Limit `Key` to string or number
+			Key extends string | number // Limit `Key` to string or number.
 				? T[Key] extends UnknownRecord | UnknownArray
 					? (
 						IsNever<Paths<T[Key]>> extends false
-						// If `Key` is number, return `Key | `${Key}``. Because both `array[0]` and `array['0']` are work.
+							// If `Key` is a number, return `Key | `${Key}``, because both `array[0]` and `array['0']` do not work.
```
Comment (Emiyaaaaa) [inline source/paths.d.ts:102]: Is this a mistake? @sindresorhus

Comment (sindresorhus) [reply source/paths.d.ts:102]: Yes, fixed.

**Pull request merged**